vsintegration/src/FSharp.Compiler.Server/Ipc/IpcClient.fs

```diff
@@ -0,0 +1,91 @@
+﻿namespace Microsoft.FSharp.Compiler.Server
+
+open System
+open System.IO
+open System.Diagnostics
+open System.Threading
+open System.Threading.Tasks
+open System.IO.Pipes
+
+open Newtonsoft.Json
+
+type IpcMessage<'Send, 'Receive> =
+    {
+        Data: 'Send
+        Reply: AsyncReplyChannel<Result<'Receive, string>>
+    }
+
+type IpcMessageClient<'Send, 'Receive>(name) =
+
+    let buffer = Array.zeroCreate<char> Constants.IpcBufferSize
+
+    let restartingEvent = Event<unit>()
+
+    let mutable agent = Unchecked.defaultof<_>
```

Comment:
Better to make invalid states impossible, and an IpcMessageClient is constructed in the started state.